src/main/scala/shine/OpenCL/CodeGeneration/InsertMemoryBarriers.scala

```diff
@@ -0,0 +1,208 @@
+package shine.OpenCL.CodeGeneration
+
+import shine.DPIA.ImperativePrimitives._
+import shine.DPIA.Phrases.{VisitAndRebuild, _}
+import shine.DPIA.Types._
+import shine.DPIA._
+import shine.OpenCL.ImperativePrimitives._
+import shine._
+
+import scala.collection.mutable
+
+object InsertMemoryBarriers {
+  def apply(p: Phrase[CommType]): Phrase[CommType] = {
+    val (p2, m) = visitNewMetadata(p, Map())
+    p2
+  }
+
+  private case class Metadata(// reads from outer scope allocations
+                              reads: mutable.Map[Identifier[_ <: PhraseType], AddressSpace],
+                              // work-group parallel writes to outer scope allocations
+                              wg_writes: mutable.Map[Identifier[_ <: PhraseType], AddressSpace])
+
+  private def visitNewMetadata(p: Phrase[CommType],
```

Comment:
The name of this function is generic and does not seem to suggest what is going on inside.

Reply:
I could call it `metadataOfCommand`, or `computeMetadata`.

Reply:
Hmm well it also modifies the command (inserting barriers), so maybe `visitAndGetMetadata`.

Reply:
Exactly, isn't this just `analyzeAndInsertBarrier` or something like this?

Reply:
sounds like a good name, I am changing it